src/module.c

```diff
@@ -675,6 +675,7 @@ void moduleFreeContext(RedisModuleCtx *ctx) {
                 propagatePendingCommands();
             if (server.busy_module_yield_flags) {
                 blockingOperationEnds();
+                unprotectClient(ctx->client);
```

Comment:
i now realize, that this could be a different client than the one we used for protectClient.
i.e. in case there are nested RM_Call or alike, the unprotect will be called on the root client (when the last nesting level is cleared), but the protect can be called on another client, done by one of the inner contexts.

@MeirShpilraien am i right? do you have a suggestion?

Reply:
>i now realize, that this could be a different client than the one we used for protectClient.

@oranagra it's not right, the root cause is `server.busy_module_yield_flags` is wrong, because of the codes:
```c
void RM_Yield(RedisModuleCtx *ctx, int flags, const char *busy_reply) {
    ...
            server.busy_module_yield_reply = prev_busy_module_yield_reply;
            /* Possibly restore the previous flags in case of two nested contexts
             * that use this API with different flags, but keep the first bit
             * (PROCESS_EVENTS) set, so we know to call blockingOperationEnds on time. */
            server.busy_module_yield_flags &= ~BUSY_MODULE_YIELD_CLIENTS;
```
I don't understand why we do like this, IMHO `RM_Yield` should not be called nested.

The `RM_Yield` API is introduced in #9963, @perryitay can you provide some clues?

Reply:
the unprotecting client is the same client protected in `RM_Yield`, but the `unproteClient` (and other codes in server.busy_module_yield_flags condition scope) would never be called in `moduleFreeContext`, since the `server.busy_module_yield_flags` is wrong.

Reply:
>we decided to allow RM_Yield from nested contexts and attempted to handle things properly.

could you give me some examples? I didn't follow the whole original PR.

Reply:
imagine a client executing a module command, and that module command calls RM_Call that's a command of another module (and so on). some of these module commands can call RM_Yield.
the client in the context of each of these will not be the root client that executed the command, but rather a temporary client the outer module used for the RM_Call.
the first call to RM_Yield should start the blocking operation (and protect the real client), and the whole thing should be undone only when the outer most context is released.

> the unprotecting client is the same client protected in RM_Yield, 

i'm still missing something. it doesn't look right to me.

> the unproteClient (and other codes in server.busy_module_yield_flags condition scope) would never be called in moduleFreeContext, since the server.busy_module_yield_flags is wrong.

i don't see the bug, could you point me to it?

Reply:
this bug is very easy to reproduce, test case `tests/modules/blockedclient.c` is the example:

1. call `RedisModule_Yield` with `REDISMODULE_YIELD_FLAG_CLIENTS` flag
    ```c
    int slow_fg_command(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
        ...
        int yield_flags = block_time? REDISMODULE_YIELD_FLAG_NONE: REDISMODULE_YIELD_FLAG_CLIENTS;
        while (!abort_flag) {
            RedisModule_Yield(ctx, yield_flags, "Slow module operation");
            usleep(1000);
            if (block_time && RedisModule_MonotonicMicroseconds() - start_time > (uint64_t)block_time)
                break;
        }
    ```
2. in module.c, `RM_Yield` set `server.busy_module_yield_flags` with `BUSY_MODULE_YIELD_CLIENTS`
    ```c
    void RM_Yield(RedisModuleCtx *ctx, int flags, const char *busy_reply) {
        ...
            if (!server.busy_module_yield_flags) {
                server.busy_module_yield_flags = flags & REDISMODULE_YIELD_FLAG_CLIENTS ?
                    BUSY_MODULE_YIELD_CLIENTS : BUSY_MODULE_YIELD_EVENTS;
                blockingOperationStarts();
                protectClient(ctx->client);
            }
    ```
3. but then `RM_Yield` remove the `BUSY_MODULE_YIELD_CLIENTS` flag, so `server.busy_module_yield_flags` is 0
    ```c
            processEventsWhileBlocked();

            server.busy_module_yield_reply = prev_busy_module_yield_reply;
            /* Possibly restore the previous flags in case of two nested contexts
             * that use this API with different flags, but keep the first bit
             * (PROCESS_EVENTS) set, so we know to call blockingOperationEnds on time. */
            server.busy_module_yield_flags &= ~BUSY_MODULE_YIELD_CLIENTS;
    ```
4. after stop_slow_fg_command executed, `slow_fg_command` breaks and module system call `moduleFreeContext`, but `server.busy_module_yield_flags` is 0, so some codes cannot be executed
    ```c
    void moduleFreeContext(RedisModuleCtx *ctx) {
            if (server.busy_module_yield_flags) {
                blockingOperationEnds();
                unprotectClient(ctx->client);
                server.busy_module_yield_flags = BUSY_MODULE_YIELD_NONE;
                unblockPostponedClients();
            }    
    ```

Reply:
>imagine a client executing a module command, and that module command calls RM_Call that's a command of another module (and so on). some of these module commands can call RM_Yield.
the client in the context of each of these will not be the root client that executed the command, but rather a temporary client the outer module used for the RM_Call.
the first call to RM_Yield should start the blocking operation (and protect the real client), and the whole thing should be undone only when the outer most context is released.

I know this case, the **nested** I mean here is some module command call `RM_Yield` in `RM_Yield`'s `processEventsWhileBlocked`

Reply:
>  the nested I mean here is some module command call RM_Yield in RM_Yield's processEventsWhileBlocked

no, i didn't mean RM_Yield itself is nested.. that that it is used inside nested calls to RM_Call (several different module contexts involved)

Reply:
> but then RM_Yield remove the BUSY_MODULE_YIELD_CLIENTS flag, so server.busy_module_yield_flags is 0

ohh, now i see it.
i think this is the bug (fix):
```diff
-                server.busy_module_yield_flags = flags & REDISMODULE_YIELD_FLAG_CLIENTS ?
-                    BUSY_MODULE_YIELD_CLIENTS : BUSY_MODULE_YIELD_EVENTS;
+                server.busy_module_yield_flags = (flags & REDISMODULE_YIELD_FLAG_CLIENTS ?
+                    BUSY_MODULE_YIELD_CLIENTS : 0) | BUSY_MODULE_YIELD_EVENTS;
```

i.e. I meant to always set `BUSY_MODULE_YIELD_EVENTS`, and sometimes also set BUSY_MODULE_YIELD_CLIENTS

Reply:
so considering this bug will be fixed, what i'm still missing is in the case of nested RM_Calls, and when RM_Yield is used on one of the inner contexts. how can we make sure to protect / unprotect the right client...

Reply:
@oranagra I believe you are right, you might protect and unprotected different clients.

Reply:
>no, i didn't mean RM_Yield itself is nested.. that that it is used inside nested calls to RM_Call (several different module contexts involved)

I think we should add a check about `RM_Yield` itself nested, module is too flexible, imagine user implement a `allow-busy` command and call `RM_Yield`.

>so considering this bug will be fixed, what i'm still missing is in the case of nested RM_Calls, and when RM_Yield is used on one of the inner contexts. how can we make sure to protect / unprotect the right client...

Yes, I think we need another PR to fix it, the bug is not a part of this PR, I can rebase after bug fixed. Maybe add a new global variable `server.root_` like `server.lua_caller`?

Reply:
#10548